### PR TITLE
🧠 discord abomination

### PR DESCRIPTION
**Pull request by joelhooks**
This sets up basic membership syncing from egghead-next to Discord.

### CODE REVIEW

File: .env.development [hunk 1]
```diff
@@ -4,3 +4,10 @@ NEXT_PUBLIC_REDIRECT_URI=https://next.egghead.io
 NEXT_PUBLIC_DEPLOYMENT_URL=http://localhost:3000
 NEXT_PUBLIC_ALGOLIA_KEY=d1b3f68acf6c2817900630bc0ac6c389
 NEXT_PUBLIC_ALGOLIA_APP=78FD8NWNJK
+NEXT_PUBLIC_DISCORD_AUTHORIZE=https://discord.com/api/oauth2/authorize?client_id=417792875355701249&redirect_uri=http%3A%2F%2Flocalhost%3A3000%2Fdiscord%2Fcallback&response_type=code&scope=identify%20email%20guilds%20guilds.join
```
Comment (joelhooks) [inline .env.development:7]: two secrets are required for this to work, the `DISCORD_CLIENT_SECRET` and the `DISCORD_BOT_TOKEN` - the rest of these are fine public.

File: src/components/login-required.tsx [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import React, {FunctionComponent} from 'react'
+import {useRouter} from 'next/router'
+
+export type LoginRequiredParams = {
+  loginRequired: boolean
+}
+
+const LoginRequired: FunctionComponent<LoginRequiredParams> = ({
```
Comment (joelhooks) [inline src/components/login-required.tsx:8]: this redirects to login if the param is `true`

not sure how robust it is, but it works for now.

Comment (joelhooks) [reply src/components/login-required.tsx:8]: mostly it just prevents the user from seeing a broken state so if they haXx0r it's on them

File: src/pages/api/discord.ts [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+    throw new Error('no discord guild was found')
+  }
+}
+
+export default async (req: NextApiRequest, res: NextApiResponse) => {
+  if (req.method === 'POST') {
+    try {
+      const client = await getDiscordBotClient()
+      const {discordUser, discordToken} = await loadDiscordUser(req.body.code)
+      const {eggheadToken} = getTokenFromCookieHeaders(
```
Comment (joelhooks) [inline src/pages/api/discord.ts:181]: we store an auth token on the cookie and it's available on api/server functions 🎉

File: src/pages/api/discord.ts [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+  } else {
+    throw new Error('no discord guild was found')
+  }
+}
+
+export default async (req: NextApiRequest, res: NextApiResponse) => {
+  if (req.method === 'POST') {
+    try {
+      const client = await getDiscordBotClient()
+      const {discordUser, discordToken} = await loadDiscordUser(req.body.code)
```
Comment (joelhooks) [inline src/pages/api/discord.ts:180]: this is kicked off after the oauth flow through discord and back to the callback URL. The code is a query param sent by Discord.

Comment (joelhooks) [reply src/pages/api/discord.ts:180]: with the code we can get the user object and a discord token to use to add them to the egghead discord

File: src/pages/api/discord.ts [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+      const {eggheadToken} = getTokenFromCookieHeaders(
+        req.headers.cookie as string,
+      )
+
+      if (discordUser && eggheadToken) {
+        const discordMember: GuildMember = await fetchOrInviteDiscordMember(
```
Comment (joelhooks) [inline src/pages/api/discord.ts:186]: a discord user is different than a server (guild - Discords game roots shine) member. We want to either fetch the member or give them an "instant" invite to the server.

File: src/pages/api/discord.ts [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+          client,
+          discordUser,
+          discordToken,
+        )
+
+        const eggheadUser: Viewer = await fetchEggheadUser(eggheadToken)
```
Comment (joelhooks) [inline src/pages/api/discord.ts:192]: now we can grab the egghead user!

File: src/pages/api/discord.ts [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+
+        await updateDiscordRolesForEggheadUser({
```
Comment (joelhooks) [inline src/pages/api/discord.ts:194]: now we will check the egghead user status and add a `member` role if they qualify

**Comment by joelhooks** [comment]
I put the appropriate keys into dev/prod environments in vercel project. Should be good to go.

File: src/pages/api/discord.ts (hunk 3 above)
Comment (johnlindquist) [reply src/pages/api/discord.ts:181]: ♥️ the api route. I didn't realize how valuable the same-domain serverless fns concept was until we started using it.

File: src/pages/discord/callback.tsx [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+const CallbackPage: FunctionComponent<LoginRequiredParams> = ({
+  loginRequired,
+}) => {
+  const [syncingAccount, setSyncingAccount] = React.useState(true)
+  const router = useRouter()
+
+  React.useEffect(() => {
+    async function connectUser() {
+      const queryHash = queryString.parse(window.location.search)
```
Comment (johnlindquist) [inline src/pages/discord/callback.tsx:19]: I think `router.query.code` elimates the need for `queryString` and `window`

**Review by johnlindquist — APPROVED** [review]

**Pull request merged**